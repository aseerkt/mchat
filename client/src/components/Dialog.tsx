import { useRef } from 'react'
import { createPortal } from 'react-dom'
import { useClickOutside } from '../hooks/useClickOutside'
import { cn } from '../utils/style'

interface DialogProps {
  isOpen: boolean
  onClose: () => void
  children: React.ReactNode
}

export const Dialog = ({ isOpen, onClose, children }: DialogProps) => {
  const dialogRef = useRef<HTMLDivElement>(null)

  useClickOutside(dialogRef, onClose)

  return createPortal(
    <div
      aria-label='dialog_overlay'
      className={cn(
        'fixed flex h-screen w-screen items-center justify-center',
        isOpen && 'hidden',
      )}
    >
      <div ref={dialogRef} role='dialog' aria-modal='true'>
        {children}
      </div>
    </div>,
    document.body,
  )
}
