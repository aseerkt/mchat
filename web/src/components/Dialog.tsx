import { useRef } from 'react'
import { createPortal } from 'react-dom'
import { useClickOutside } from '../hooks/useClickOutside'
import { cn } from '../utils/style'

interface DialogProps {
  isOpen: boolean
  onClose: () => void
  children: React.ReactNode
}

export const Dialog = ({ isOpen, onClose, children }: DialogProps) => {
  const dialogRef = useRef<HTMLDivElement>(null)

  useClickOutside(dialogRef, onClose)

  return isOpen
    ? createPortal(
        <div
          aria-label='dialog_overlay '
          className={cn(
            'fixed top-0 z-20 flex h-screen w-screen items-center justify-center bg-black bg-opacity-80',
            !isOpen && 'hidden',
          )}
        >
          <div
            ref={dialogRef}
            role='dialog'
            aria-modal='true'
            className='border-3 w-full max-w-[500px] shrink-0 rounded bg-white p-6 shadow-md'
          >
            {children}
          </div>
        </div>,
        document.body,
      )
    : null
}
