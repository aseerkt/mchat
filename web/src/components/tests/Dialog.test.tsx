import { render } from '@testing-library/react'
import { Dialog } from '../Dialog'

describe('<Dialog /> tests', () => {
  it('should match snapshot for isOpen false', () => {
    const { baseElement } = render(
      <Dialog isOpen={false} onClose={() => {}}>
        hello world
      </Dialog>,
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should match snapshot for isOpen true', () => {
    const { baseElement } = render(
      <Dialog isOpen={true} onClose={() => {}}>
        hello world
      </Dialog>,
    )
    expect(baseElement).toMatchSnapshot()
  })
})
