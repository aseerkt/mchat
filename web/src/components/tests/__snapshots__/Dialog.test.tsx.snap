// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Dialog /> tests > should match snapshot for isOpen false 1`] = `
<body>
  <div />
</body>
`;

exports[`<Dialog /> tests > should match snapshot for isOpen false 2`] = `
<body>
  <div />
  <div
    aria-label="dialog_overlay"
    class="fixed top-0 flex h-screen w-screen items-center justify-center bg-black bg-opacity-80"
  >
    <div
      aria-modal="true"
      class="border-3 rounded bg-white p-6 shadow-md"
      role="dialog"
    >
      hello world
    </div>
  </div>
</body>
`;

exports[`<Dialog /> tests > should match snapshot for isOpen true 1`] = `
<body>
  <div />
  <div
    aria-label="dialog_overlay"
    class="fixed top-0 flex h-screen w-screen items-center justify-center bg-black bg-opacity-80"
  >
    <div
      aria-modal="true"
      class="border-3 rounded bg-white p-6 shadow-md"
      role="dialog"
    >
      hello world
    </div>
  </div>
</body>
`;
